### PR TITLE
fix regression in time-like check when decoding masked data

### LINKED ISSUE #8269: open_dataset with engine='zarr' changed from '2023.8.0' to '2023.9.0'
### What is your issue?
When moving from xarray version '2023.8.0' to '2023.9.0' the behavior of importing a zarr changed for me (code to create example zarr at end of this post). When importing a variable with units "days accumulated", the values are scaled differently between the two versions. The latest version seems to automatically treat this as as time-like array (I think the -9.223372e+18 seen are NaT-like?).

Open the zarr:
```python
import xarray as xr
ds = xr.open_dataset('debug.zarr', engine='zarr', chunks={})
```

Print as a pandas-like table for each version of xarray for readability:
```python
 ds.to_dataframe()
```

Version '2023.8.0':
|time|dapr (dtype=float32)|mdpr (dtype=float32)|
|---|---|---|
|2000-01-01|NaN|NaN|
|2000-01-02|NaN|NaN|
|2000-01-03|2.0|1.5|

Version '2023.9.0':
|time|dapr (dtype=float64)|mdpr (dtype=float32)|
|---|---|---|
|2000-01-01|-9.223372e+18|NaN|
|2000-01-02|-9.223372e+18|NaN|
|2000-01-03|2.000000e+00|1.5|

I can manually disable this by using the "use_cf=False", "mask_and_scale=False", and then manually scale this variable, though that is not ideal. The "decode_timedelta" doesn't seem to have an effect on this data, either.

I understand the "days" keyword is in my units, however the full unit is "days accumulated". Has the behavior of xarray changed to find keywords such as "days" occurring anywhere in the units (eg. as a substring)? Do you have any other suggestions? Thank you for the help.

### Code to create the debug.zarr for the tables above:

```python
import numpy as np
import pandas as pd
import xarray as xr
import zarr

# Create some multiday precipitation data (similar to https://www.ncei.noaa.gov/products/land-based-station/global-historical-climatology-network-daily)
# mdpr is the amount of a multiday total (inches)
# dapr is the number of days each multiday total occurred over (days accumulated).
# In this example, 1.50 inches of rain fell over 2 days (2 observation periods), ending on 2000-01-03
# I use float32 to represent these, but pack these as int16 values in the zarr.
mdpr = np.array([np.NaN, np.NaN, 1.50], dtype=np.float32)
dapr = np.array([np.NaN, np.NaN , 2.0], dtype=np.float32)
time = pd.date_range('2000-01-01', periods=3)

# Create a dataset from these values
ds = xr.Dataset(
    data_vars=dict(
        mdpr=(['time'], mdpr),
        dapr=(['time'], dapr),
    ),
    coords=dict(
        time=time,
    ),
    attrs=dict(description='multiday precipitation data'),
)

# Specify encoding to pack these float32 values as int16
encoding = {
    'mdpr' : {
        'chunks' : (3,),
        'compressor': zarr.Blosc(cname='zstd', clevel=3, shuffle=1),
        'filters': None,
        'missing_value': -32768,
        '_FillValue': -32768,
        'scale_factor': 0.01,
        'add_offset': 0.0,
        'dtype': np.int16,
    },
    'dapr' : {
        'chunks' : (3,),
        'compressor': zarr.Blosc(cname='zstd', clevel=3, shuffle=1),
        'filters': None,
        'missing_value': -32768,
        '_FillValue': -32768,
        'scale_factor': 1.0,
        'add_offset': 0.0,
        'dtype': np.int16,
    },
}

# Create attributes. The "units" for the dapr variable seems to be the issue "days" in the
# "days accumulated"
ds.mdpr.attrs['units'] = 'inches'
ds.mdpr.attrs['description'] = 'multiday precip amount'

ds.dapr.attrs['units'] = 'days accumulated'
ds.dapr.attrs['description'] = 'number of days included in the multiday precipitation'

# Save to zarr
ds.to_zarr('debug.zarr', mode='w', encoding=encoding)
```

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] Closes #8269
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
This fix works for timedelta `units` in any case by checking for equality with any of the unit-strings (e.g. `days`) For datetime `unit` presence of `since` is checked in addition to presence of any of the unit-strings (e.g. `nanoseconds`).

There might be other unit-strings possible, like `days since system setup` which still will be detected as time-like. To even catch these cases we would need to parse the whole units string for the three parts (`"UNIT since DATE"`).

**Comment by spencerkclark** [comment]
Thanks for tracking this down @kmuehlbauer -- looks good.  Could you maybe just add a simple test?

> There might be other unit-strings possible, like `days since system setup` which still will be detected as time-like. To even catch these cases we would need to parse the whole units string for the three parts (`"UNIT since DATE"`).

Right, yeah, I think this is something we have been living with for a while within our datetime decoding logic: https://github.com/pydata/xarray/blob/2cd8f96a1b5ae954d7b34390e8b01fbd985fc710/xarray/coding/times.py#L826

In that case (regardless of what happens during masking) an error will eventually be raised related to the fact that what comes after `"since"` is not a valid date, which I think is OK.

**Comment by kmuehlbauer** [comment]
Thanks @spencerkclark, I'll add a test along the lines of the example in #8269 and check that variables with non time-like units are not treated as time-like.

**Comment by spencerkclark** [comment]
Sounds good, thanks!

> In that case (regardless of what happens during masking) an error will eventually be raised related to the fact that what comes after "since" is not a valid date, which I think is OK.

Hmm as I think about this more, I guess we might need to worry about the case when one uses `decode_times=False` when opening a file.  In that case what happens during masking could matter, so maybe we do need some more careful checking there after all.  I think running the units through `coding.times._unpack_netcdf_time_units` and ensuring that no error is raised might be sufficient.

**Comment by spencerkclark** [comment]
Maybe we actually take things a step further and not apply this special time-masking behavior at all in the case that `decode_times` or `decode_timedelta` are `False`.  Similar to the `use_cftime` parameter, we could add those parameters to the constructor for the `CFMaskCoder` and use them during decoding.  

What are your thoughts on that?  I guess I'm wondering whether this special masking behavior (replacing integer missing values with the minimum integer) should apply if we are not ultimately decoding values as times.

**Comment by kmuehlbauer** [comment]
I've wrapped my head around that many times without coming to a robust and nice solution. Sometimes we can't clearly separate the different coding steps. 

Consider the following example. If we do not keep/treat time-like with int64 resolution in `CFMaskCoder` it's not possible to do something like this (with packed data):

```python
ds = xr.open_dataset(filename, decode_times=False)
ds = ds.pipe(fix_time_units)
ds = xr.decode_cf(decode_times=True)
```
without losing nanosecond resolution (because conversion into floating point).

Yes, we could advertise to complete drop cf decoding in those cases instead of just dropping time decoding.

**Comment by spencerkclark** [comment]
Good point, I guess it really is a coupled problem...one way or the other you will need to completely drop CF decoding to fix something (either remove the time-like units to prevent special decoding of time-like missing values, or fix the time units without losing nanosecond precision).

On balance as a user I might be more surprised to see time-like units playing a role in how fields were masked in the case that `decode_times=False` or `decode_timedelta=False` than if nanosecond precision was lost after fixing the time units.  I'm open to opinions from you and others though -- maybe @dcherian also has thoughts?

**Comment by spencerkclark** [comment]
Maybe we can punt on the question of whether to apply the special time masking or not when `decode_times` or `decode_timedelta` are `False`, since it is not required to answer to address #8269?  It would be good to get at least this fix in.

> We might have to move that functionality over to `coding.variables` to not run into circular dependencies. Or is there another solution to this.

Ah gotcha, for this, one alternative is to import `coding.times._unpack_netcdf_time_units` within the `_is_time_like` function.

**Comment by kmuehlbauer** [comment]
Thanks Spencer! I'm currently traveling and can pick up work next week. Feel free to push this forward.

File: xarray/tests/test_conventions.py [hunk 1]
```diff
@@ -333,6 +333,16 @@ def test_invalid_time_units_raises_eagerly(self) -> None:
...
 
+    def test_invalid_timedelta_units_do_not_decode(self) -> None:
+        # regression test for #8269
+        ds = Dataset(
+            {"time": ("time", [0, 1, 20], {"units": "days invalid", "_FillValue": 20})}
+        )
+        expected = Dataset(
+            {"time": ("time", [0.0, 1.0, np.nan], {"units": "days invalid"})}
+        )
+        assert_identical(expected, decode_cf(ds))
```
Comment (spencerkclark) [inline xarray/tests/test_conventions.py:344]: Could you parametrize this test to make sure this works for both the `decode_times=True` and `decode_times=False` cases?

**Review by spencerkclark — APPROVED** [review]
Thanks @kmuehlbauer!

**Pull request merged**